[Config] Renamings to allow all callback types

mpajunen opened the pull request:
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

The function and property name changes in preparation of #14330 to allow using any callable with config builders.

mpajunen commented:
ping @stof (or someone else), any more thoughts on this?

nicolas-grekas commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -25,8 +25,8 @@
-    protected $normalizationClosures = array();
-    protected $finalValidationClosures = array();
+    protected $normalizationCallbacks = array();
+    protected $finalValidationCallbacks = array();
```
Aliasing the old ones by reference would allow preserving BC.
you should do it in the constructor, and for completeness also add a magic __clone method:
```php

public function __construct(...)
{
    $this->normalizationClosures = &$this->normalizationCallbacks;
    $this->finalValidationClosures = &$this->finalValidationCallbacks;

    // ...
}
// ...
public function __clone()
{
    // Break cross-clones references but preserve them inside each one.
    unset($this->normalizationCallbacks, $this->finalValidationCallbacks);
    $this->normalizationCallbacks = $this->normalizationClosures;
    $this->finalValidationCallbacks = $this->finalValidationClosures;
    $this->normalizationClosures = &$this->normalizationCallbacks;
    $this->finalValidationClosures = &$this->finalValidationCallbacks;
}
```

mpajunen replied:
Thanks. Referenced aliases added.

nicolas-grekas replied:
You forgot to keep the declaration above

mpajunen replied:
Works without, but of course it makes sense to keep the old property declarations as well. Restored now, with deprecation notes.

GromNaN commented:
What is the benefit of renaming the method and property ?

mpajunen commented:
There is no immediate benefit, this PR is simply a preparation step for #14330 that allows any `callable` to be used instead of just `\Closure`s. The current names would be confusing then.

nicolas-grekas commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -152,23 +164,51 @@ public function setAllowOverwrite($allow)
     public function setNormalizationClosures(array $closures)
     {
-        $this->normalizationClosures = $closures;
+        trigger_error('The '.__METHOD__.' method is deprecated since version 2.8 and will be removed in 3.0. Use the setNormalizationCallbacks() method instead.', E_USER_DEPRECATED);
```
Please add `@` in front of trigger_error: `@trigger_error(...`, same below

mpajunen replied:
Done.

mpajunen committed:
[Config] Use callable instead of closure in naming

Replace references to closures in function, property and
parameter names in preparation of future callable support.

xabbuh commented on src/Symfony/Component/Config/Definition/BaseNode.php:
```diff
@@ -48,6 +56,10 @@ public function __construct($name, NodeInterface $parent = null)
+
+        // Backwards compatibility for old property names, to be removed in 3.0
+        $this->normalizationClosures = &$this->normalizationCallbacks;
+        $this->finalValidationClosures = &$this->finalValidationCallbacks;
```
Would it be an option to remove the `normalizationCallbacks` and `finalValidationCallbacks` properties and implementing `__get()` and `__set()` for them instead? This way we could also trigger deprecation notices when the deprecated properties are used.

mpajunen replied:
Yes, using magic methods would allow allow deprecation notices and also somewhat simpler implementation (no need for `__clone()` tricks). It would also make the properties effectively public, which could be a problem. In addition the new properties could also probably be private, unless there's some specific reason the old ones were protected.

All of these changes look like bikeshedding to me though. I checked with Packanalyst earlier and found just one class that would need to be changed for 3.0 compatibility -- and that one uses the setter methods. A quick search is not going to find every use case, but I think using the properties directly is likely rare even in custom subclasses.

Of course, I can still make the changes if there's agreement that using `__get()` and `__set()` would be preferred.

fabpot commented:
Closing this PR as the other one was closed as well. Supporting closure only in this case seems enough.
